CDCgov/prime-reportstream
# Experience/2246: Clean CSV Data: Clear Illness Onset Date Field when Date is Invalid

This PR introduces a mapper that nullifies dates that to do not conform to the specified/expected date format.

https://app.zenhub.com/workspaces/prime-reportstream-5ff4833beb3e08001a4cacae/issues/cdcgov/prime-reportstream/2246

**If you are suggesting a fix for a currently exploitable issue, please disclose the issue to the prime-reportstream team directly outside of GitHub instead of filing a PR, so we may immediately patch the affected systems before a disclosure. See [SECURITY.md/Reporting a Vulnerability](https://github.com/CDCgov/prime-reportstream/blob/master/SECURITY.md#reporting-a-vulnerability) for more information.**

Test Steps:
1. include the new mapper in a schema element that is a DATE type 
ex:
```
- name: test_result_date
  csvFields: [{ name: testResultDate}]
  mapper: nullDateValidator($dateFormat:MM/dd/yyyy, test_result_date)
```
2. update the data that uses the schema to have some invalid dates in that field, like "a week ago" or "202106" or something like that 
3. send data to RS using that schema
4. the value for the date should be an empty string

## Changes
- added `nullDateValidator` mapper
- added `$dateFormat:` mapper token
- added unit tests for token parsing
- added unit tests for `nullDateValidator` mapper
- added unit tests for tokenized value mapping

## Checklist

### Testing

### Process

## Fixes
- #issue - List GitHub issues this PR fixes

## To Be Done
*Create GitHub issues to track the work remaining, if any*
- #issue

## Specific Security-related subjects a reviewer should pay specific attention to
- Does this PR introduce new endpoints?
    - new endpoint A
    - new endpoint B
- Does this PR include changes in authentication and/or authorization of existing endpoints?
- Does this change introduce new dependencies that need vetting?
- Does this change require changes to our infrastructure?
- Does logging contain sensitive data?
- Does this PR include or remove any sensitive information itself?

If you answered '_yes_' to any of the questions above, conduct a detailed Review that addresses at least:

- What are the potential security threats and mitigations? Please list the _STRIDE_ threats and how they are mitigated
    - **S**poofing (faking authenticity)
        - Threat _T_, which could be achieved by _A_, is mitigated by _M_
    - **T**ampering (influence or sabotage the integrity of information, data, or system)
    - **R**epudiation (the ability to dispute the origin or originator of an action)
    - **I**nformation disclosure (data made available to entities who should not have it)
    - **D**enial of service (make a resource unavailable)
    - **E**levation of Privilege (reduce restrictions that apply or gain privileges one should not have)
- Have you ensured logging does not contain sensitive data?
- Have you received any additional approvals needed for this change?
